### ISSUE TITLE
Make leojs files convert ua's to stringified json when objects are compatible. and omit un-jsonable objects.

### ISSUE THREAD

**Issue by boltex** [leo-editor/leo-editor#2666]
**Background:** While trying to support pickling and unpickling ua's when dealing with ".leo" files while i'm coding leojs (the vscode extension, not the file format), I thought about checking what Edward wrote when Leo writes ".leojs" files to disk for inspiration, because it's the exact opposite as what i'm trying to do... : The ua's are ignored when writing the file for now. 

So i thought i might add this functionality to Leo, and I considered having the ".leojs" file format maybe keep ua's as pickled + hexlified strings, kept as-is like in Leo, **but it seems more elegant and simple to me to just include the ua's in the native javascript format: JSON**.

Changed plans : just add a ua dict at the end with gnx as keys.
~~So, i propose that I add functionality to Leo, for T nodes of ".leojs" files, to simply have ua's as stringified json. which is trivial to implement. And, mirroring this write operation, to add support for reading ua's from leojs file while opening such a file.~~

~~**Pros**: human readable, easily readable and writable in both Leo and LeoJs. (and other future misc. language implementations)~~
~~**Cons**: Loss of some format info (sets may be brought back as simple lists, no class or function, instances)~~

Obviously, while doing so, a warning should be given to the user if an un-jsonable (un-serializable object) is encountered, while nonetheless continue on serializing the rest, and produce a .leojs file.

This would allow people who want a 'JSON / web api ' compatible file and data format to have it, by using .leojs files. 

People who use the typical Leo workflow, with regular xml ".leo" files + python pickled ua objects, are not affected at all by those new ".leojs" read/write methods.

(I will obviously mirror those behaviors in the leojs vscode expansion for both .leojs files using JSON, and by using a special library to unencode and reencode hexlify, pickled ua's in regular xml ".leo" files)

Félix

**Comment by edreamleo** [comment]
@boltex This proposal seems reasonable, but what is a T node ???

**Comment by boltex** [comment]
@edreamleo 

T nodes was just my ad-hoc way of talking about the equivalent of t tag elements, which are child of the tnodes element.

Leo files have "v tags" and "t tags" elements, whereas ".leojs" have a vnodes array, and a tnodes dict. And i now realize that "t nodes' don't exist in .leojs files as the tnodes dict can only provide for body content. So ua's would have to be in another dict, or, have the tnodes dict be able to contain other dicts instead of just the direct string, like so (note the usage of the 'tx' attribute name, which fits perfectly as not available for ua name anyways,): 

```
{
  "leoHeader": {
    "fileFormat": 2
  },
  "globals": {},
  "tnodes": {
    "felix.20201116231334.2": "case 1, I'm just body text",
    "felix.20201116231336.1": { "tx": "case 2, Body text in tx attribute",  "foo": "some ua", "bar": "{\"name\":\"John\",\"age\":30,\"myArray\":[1,2,\"a string\"]}" }
 },
"vnodes": [...]
}

```

**Comment by edreamleo** [comment]
@boltex Thanks for the clarifications. Leo used to have tnodes, and yes, they corresponded (roughly) to <t> elements. tnodes went away many years ago with the advent of the so-called "one-node" world.

I don't understand all the implications of your comment, but I'm pretty sure you do, hehe. Keep up the great work.

**Comment by boltex** [comment]
I fixed it #2666 along with #2682 in felixserver4 - I simply put the ua's in an optional ua dict. So much more effective and simple, dont know why i didnt think of this in the first place.

I'll close this issue when this gets reviewed & merged.

**Comment by boltex** [comment]
Closing this as it's now part of devel - from the felixserver4 branch :)
